BraintreeCore/src/main/java/com/braintreepayments/api/AuthorizationLoader.java

```diff
@@ -0,0 +1,59 @@
+package com.braintreepayments.api;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+class AuthorizationLoader {
+
+    private Authorization authorization;
+    private final ClientTokenProvider clientTokenProvider;
+
+    AuthorizationLoader(@Nullable String initialAuthString, @Nullable ClientTokenProvider clientTokenProvider) {
+        this.clientTokenProvider = clientTokenProvider;
+        if (initialAuthString != null) {
+            this.authorization = Authorization.fromString(initialAuthString);
+        }
+    }
+
+    void loadAuthorization(@NonNull final AuthorizationCallback callback) {
+        if (authorization != null) {
+            callback.onAuthorizationResult(authorization, null);
+        } else if (clientTokenProvider != null) {
+            clientTokenProvider.getClientToken(new ClientTokenCallback() {
+                @Override
+                public void onSuccess(@NonNull String clientToken) {
+                    authorization = Authorization.fromString(clientToken);
+                    callback.onAuthorizationResult(authorization, null);
+                }
+
+                @Override
+                public void onFailure(@NonNull Exception exception) {
+                    callback.onAuthorizationResult(null, exception);
+                }
+            });
+        } else {
+            callback.onAuthorizationResult(null, new BraintreeException("Unable to fetch client token"));
+        }
+    }
+
+    @Nullable
+    Authorization getAuthorizationFromCache() {
+        return authorization;
+    }
+
+    AuthorizationType getAuthorizationType() {
+        if (authorization instanceof InvalidAuthorization) {
+            return AuthorizationType.INVALID;
+        }
+
+        if (authorization instanceof TokenizationKey) {
+            return AuthorizationType.TOKENIZATION_KEY;
+        }
+
+        if (clientTokenProvider != null) {
```

Comment:
Do we not need to also confirm that it's instance of `ClientToken`? Might be good to check in case there is a different auth type in the future

Reply:
For here we're assuming if the merchant has implemented a `ClientTokenProvider`, we can infer that they will fetch a `ClientToken` without having to fetch the token before it's needed. That is a good point for future auth types, I think there may be a way to remove the need for this method.